[REVIEW] Define and implement new string capitalize and title API

aschaffer opened the pull request:
This PR tracks work on string capitalize (https://github.com/rapidsai/cudf/issues/3995), but it also presents a common header for `title` and `wrap` APIs (https://github.com/rapidsai/cudf/issues/3996 and https://github.com/rapidsai/cudf/issues/3997, respectively).

aschaffer commented:
Common API header for `capitalize`, `title`, `wrap`: `include/cudf/strings/string_ops.hpp`. After we agree on the APIs there would be separate PRs for `title` and `wrap`.

davidwendt commented:
The API declarations look good. I suggest a single header file for `title()` and `capitalize()` called `capitalize.hpp` and a separate header for `wrap()` called `wrap.hpp` (in a separate PR as you mentioned).

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,356 @@
+           bytes += detail::bytes_in_char_utf8(the_chr);
+         }
+         else if( (bcapnext && IS_LOWER(flag)) || (!bcapnext && IS_UPPER(flag)) ) {
+           bcapnext = false;
```
The logic here needs to match logic for the execute pass.
Here the `bcapnext` is not being set to `false` in the else statement.
This means it will calculate the incorrect size for the chars column and possibly overrun the output buffer.
Add the following string to the Title testcase to see the error: `"SȺȺnich"`

It would be good if the probe and execute functions could be factored together to avoid this kind of coding error.

aschaffer replied:
Fixed `bcapnext` issue.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    __device__
+    virtual int32_t operator()(size_type idx) = 0;
```
This is never used. Also, it will likely not work here since the virtual-function-table needs to be created in device memory and therefore requires a class of this type be instantiated in a kernel somehow. Because the base class is never referenced in this code, the virtual function never has to be resolved.
I recommend removing this line.

aschaffer replied:
It's not _meant_ to be used. It _cannot_ be used, because it's pure virtual. The base class is an abstract class. It's only purpose is to enforce an interface to adhere to. 

The interface specifies how the derived classes are to be constructed and is a common placeholder for the private data members. Instead of duplicating code all over the place, such a base class provides a unique placeholder for the members.

Not providing an interface means having no mechanism to enforce derived classes to adhere to a common interface. They might as well be independent classes. But if they are independent, then the data members must be duplicated all over the place. So to enforce an interface to adhere to you need an abstract class. Which has zero-cost as it _cannot_ be instantiated. Hence the pure virtual function to override. Non-virtual functions _with the same name_ of classes driving from the same base are introducing more problems than they can solve. ("Item 37:  Never redefine an inherited nonvirtual function." Scott Meyers, Effective C++, 2nd ed.; since they're doing different things in the derived classes...they must be redefined; hence they must be virtual!)

Also, there is no runtime cost for this setup, because the virtual table is not used at runtime. And that is because the member function in question is _not_ called via a base pointer or reference. It's called directly via the derived object, so the compiler is optimizing the call.

Abstract classes are the _de facto_ mechanism of enforcing interfaces in C++. As long as no runtime cost is incurred (and, as explained above, it is not) there shouldn't be anything wrong with using C++ abstract classes. After all ...this is a C++ library.

davidwendt replied:
I disagree with this. If the programmer does not provide an `operator()` method then this code would not compile without or without this virtual function definition. I feel you are misleading the class user by implying polymorphism which if acted upon would fail at runtime and not compile time.

jrhemstad replied:
I understand @aschaffer's desire to want to declare the expected interface in the `execute_base` class. I agree that the way it is being used is simply to define an interface and there is no dynamic polymorphism occurring and is therefore safe and correct. 

I am not sure if it is actually _necessary_ to enforce an interface here. If someone were to attempt to use a type that derives from `execute_base` as a function object, then it would _already_ be required to have an `operator()` defined. This in effect already enforces an interface for all derived types.  

I also share @davidwendt's concern by making it pure virtual it could mislead future developers into thinking they could use a `execute_base` as a base class for dynamic polymorphism. 

> Abstract classes are the de facto mechanism of enforcing interfaces in C++. 

Perhaps, but they are not the _only_ way to enforce an interface. You could also use CRTP.

aschaffer replied:
The users have no reason to assume polymorphism, when no polymorphic container is involved. But they have the option of doing so, if they want, by involving such a container.

Assume, for a moment, that at some point one would want to apply a `vector` of such functors to a string; or, more realistically, to each row in a `string_view`.

So, one might use (semi-pseudocode-ish):
```
device_vector<probe_base*> functor_v{new probe_capitalize(...), new probe_title(...)}; 
//apply the sequence of functors to string:
for(auto const& fctr: functor_v) {
   fctr(string);
}
```
(assume for a moment that the `operator()` acts directly on `string`'s).

Do you think this code would work _without_ declaring `operator()` _virtual_? (which requires a _pure virtual_ declaration in the base class; which requires an _abstract_ class; and `overridden` versions in derived classes).

The answer is: no, it would not. Reason? Because you have non-virtual classes with same name and signature in derived classes. What you want here is dynamic polymorphism. What you get...is anything but!

**That being said, I will make the change according to your demands. But considered yourselves warned this opens a door for future disasters, in a situation like this. ;-)**

aschaffer replied:
Done.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    __device__
+    virtual int32_t operator()(size_type idx) const = 0;
```
Recommend removing this line. Detailed comment in the same method for the other base class.

aschaffer replied:
Same counter-argument as above.

aschaffer replied:
Done.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,356 @@
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+         uint32_t code_point = detail::utf8_to_codepoint(the_chr);
+         detail::character_flags_table_type flag = code_point <= 0x00FFFF ? get_flags()[code_point] : 0;
```
Perhaps this line could be made into a base-class method. The accessor method would no longer be required.
Maybe something like:
```
using char_info = thrust::pair<uint32_t,detail::character_flags_table_type>;
    __device__ char_info get_char_info(char_utf8 chr)
    {
        uint32_t code_point = detail::utf8_to_codepoint(chr);
        detail::character_flags_table_type flag = code_point <= 0x00FFFF ? d_flags[code_point] : 0;
        return char_info{code_point,flag};
    }
```

aschaffer replied:
Good idea. Done. BTW: accessor functions are also zero-cost because they are inlined and optimized away by the compiler. Nevertheless, factoring this code out is a good idea.

aschaffer replied:
Okay, replaced that everywhere.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,356 @@
+         uint32_t code_point = detail::utf8_to_codepoint(the_chr);
+         detail::character_flags_table_type flag = code_point <= 0x00FFFF ? get_flags()[code_point] : 0;
+         if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+           the_chr = detail::codepoint_to_utf8(get_case_table()[code_point]);
```
And perhaps for this line could be a base-class method as well.
Something like:
```
    __device__ char_utf8 convert_char(char_info const& info)
    {
        return detail::codepoint_to_utf8(d_case_table[info.first]);
    }
```
No need to for the `get_case_table()` method then either.

aschaffer replied:
Done.

davidwendt commented on cpp/tests/strings/case_tests.cu:
```diff
@@ -85,3 +86,35 @@ TEST_F(StringsCaseTest, EmptyStringsColumn)
+
+TEST_F(StringsCaseTest, Title)
+{
+    std::vector<const char*> h_strings{ "Examples aBc", "thesé", nullptr, "ARE THE", "tést strings", "" };
```
As mentioned in a previous comment. 
The following string will create different sized outputs: `"SȺȺnich"`
The expected string would be `"Sⱥⱥnich"`
These could be added to the `Capitalize` test as well.

aschaffer replied:
Done.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,379 @@
+         uint32_t code_point = pair_char_info.first;
+         detail::character_flags_table_type flag = pair_char_info.second;
+
+         if( !IS_ALPHA(flag) ) {
```
I recommend refactoring this to match more closely to the execute code.
```
        if( !IS_ALPHA(flag) )
          bcapnext = true;
        else
          {
            if( bcapnext ? IS_LOWER(flag) : IS_UPPER(flag) )
              the_chr = convert_char(pair_char_info);
            bcapnext = false;
          }

        bytes += detail::bytes_in_char_utf8(the_chr);
```

aschaffer replied:
Done.

aschaffer commented:
Dear reviewers, I think I addressed all review requests. Please let me know if everything okay. (especially on specific requests from @jrhemstad and @davidwendt ). Thanks.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,372 @@
+  
+  //base class for executing string modification:
+  //
+  struct execute_base
```
Could we combine the two base classes?
The `execute_base` just has two more members than `probe_base`
Perhaps the `d_offsets` and `d_chars` could be optional members
```
    capitialize_base(column_device_view const d_column,
                 character_flags_table_type const* d_flags,
                 character_cases_table_type const* d_case_table,
                 int32_t const* d_offsets = nullptr,
                 char* d_chars = nullptr ): ...
```

aschaffer replied:
Done.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,343 @@
+        return 0; // null string
+      
+      string_view d_str = get_column().template element<string_view>(idx);
+      char* d_buffer = get_chars() + get_offsets()[idx];
```
The `get_chars()` and `get_offsets()` accessors are only needed here so perhaps they could be replaced with a base class member function that simply returns the output buffer pointer. Something like:
```
    __device__ char* get_output_ptr(size_type idx)
    {
        return d_chars && d_offsets ? d_chars + d_offsets[idx] : nullptr;
    }
```
Then `get_chars()` and `get_offsets()` would no longer be needed.

aschaffer replied:
Done.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,343 @@
+  {
+    using char_info = thrust::pair<uint32_t,detail::character_flags_table_type>;
+    
+    probe_execute_base(column_device_view const d_column,
```
The `d_flags` and `d_case_table` are only needed by the base class now and can retrieve these itself rather than having the subclass pass them through. Something like this:
```
    probe_execute_base(column_device_view const d_column,
                       int32_t const* d_offsets = nullptr,
                       char* d_chars = nullptr):
      d_column_(d_column),
      d_flags_(get_character_flags_table()),      // set flag table
      d_case_table_(get_character_cases_table()), // set case table
      d_offsets_(d_offsets),
      d_chars_(d_chars)
```
Now the subclasses do not have to accept these and pass them through.
```
   execute_capitalize(column_device_view const d_column, int32_t const* d_offsets, char* d_chars)
    : probe_execute_base(d_column, d_offsets, d_chars)
    
```

aschaffer replied:
Did you mean `probe_capitalize` in your example above? Because `execute_capitalize` needs all 5 parameters to pass through.

davidwendt replied:
No. The `probe_execute_base` does not need to be passed the flag and case tables. It can retrieve them itself as shown in the example code above. Therefore `execute_capitalize` or any other subclass would no longer need to accept or to pass these through.

aschaffer replied:
I see now what you're saying. Nice. I fixed that, but I cannot check in: GitHub issues, see my entry in devops channel (internal, only)

aschaffer replied:
Okay, looks like I finally was able to check in. Resolving.

davidwendt commented on cpp/src/strings/capitalize.cu:
```diff
@@ -0,0 +1,309 @@
+
+        if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+          the_chr = convert_char(pair_char_info);
+        d_buffer += detail::from_char_utf8(the_chr, d_buffer);
```
The `probe` and `execute` functors differ by only a couple of lines of code. What about combining them into a single functor? The `d_buffer` variable is only non-null during execute so here is what capitalize could look like:
```
    __device__ int32_t operator()(size_type idx) {
      if( get_column().is_null(idx) )
        return 0; // null string

      string_view d_str = get_column().template element<string_view>(idx);
      char* d_buffer = get_output_ptr(idx);
      int32_t bytes = 0;

      for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
        auto the_chr = *itr;

        auto pair_char_info = get_char_info(the_chr); 
        detail::character_flags_table_type flag = pair_char_info.second;

        if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
          the_chr = convert_char(pair_char_info);

        if( d_buffer ) // only non-null during execute
          d_buffer += detail::from_char_utf8(the_chr, d_buffer);
        bytes += detail::bytes_in_char_utf8(the_chr);
      }
      return bytes;
    }
  };
```
This does both the probe and execute in the same function. The `bytes` are computed in both passes but is just discarded for the execute step. Or you could add an `else` clause for the `bytes += ...` line too.

This would leave one base class plus one functor for capitalize and one functor for title.
The new combined functor could have a ctor like the base class taking the offsets and chars pointers as optional parameters:
```
     probe_execute_capitalize(column_device_view const d_column,
                              int32_t const* d_offsets = nullptr,
                              char* d_chars = nullptr )
             : probe_execute_base(d_column, d_offsets, d_chars) {}
```

aschaffer replied:
I don't really see a benefit in doing so. The goal is not to decrease the number of functors at the cost of inflating the functors' operators. This would make the code hard to read and maintain. 

As stated before, merging two pieces of functionality into one breaks cohesion. The functors as they are right now have very clear responsibilities. Cocktailing them together has no clear benefits, just making things confusing for later developers maintaining or extending this functionality.

davidwendt replied:
Ok. Combining these is not a requirement.
Mostly I'm concerned someone will try to change the execute function for some feature without realizing they likely need to change the size calculation too.

aschaffer replied:
I think I fixed the sync requirement between probe and execute. Essentially, I isolated the logic, `generate_chr()` to be called inside the `operator()`'s loop so that each probe/execute pair calls the same logic for capitalize and for title, respectively. Since a picture is worth a thousand words, see the diagram attached below.

![string_capit_title](https://user-images.githubusercontent.com/37386037/75308739-ccfdd980-5814-11ea-8617-4ee8b5ac7f9f.jpg)

davidwendt replied:
This seems overly complex considering what these two simple functions do. 
I now think the original separation would be easier to follow and maintain.

aschaffer commented:
rerun tests

aschaffer commented:
rerun tests
